testsuite/tests/typing-local/local.ml

```diff
@@ -2253,3 +2253,246 @@ module F :
+       is not included in
+         type t = Bar of string
+       Constructors do not match:
+         Bar of global_ string
+       is not compatible with:
+         Bar of string
+       The first is global and the second is not.
+|}]
+
+(* global_ and nonlocal_ bind closer than star *)
+type gfoo = GFoo of global_ string * string
+type rfoo = RFoo of nonlocal_ string * string
+[%%expect{|
+type gfoo = GFoo of global_ string * string
+type rfoo = RFoo of nonlocal_ string * string
+|}]
+
+(* TESTING OF GLOBAL_ *)
+
+(* global arguments must be global when constructing
+   cannot be regional or local
+*)
+let f (local_ s : string) =
+  GFoo (s, "bar")
+[%%expect{|
+Line 2, characters 8-9:
+2 |   GFoo (s, "bar")
+            ^
+Error: This value escapes its region
+|}]
+
+let f =
+  let local_ s = "foo" in
+  GFoo (s, "bar")
+[%%expect{|
+Line 3, characters 8-9:
+3 |   GFoo (s, "bar")
+            ^
+Error: This value escapes its region
+|}]
+
+(* s' extracted from x as global *)
+(* despite x is local or regional*)
+let f (s : string) =
+  let local_ x = GFoo (s, "bar") in
+  match x with
+  | GFoo (s', _) -> ref s'
+
+[%%expect{|
+val f : string -> string ref = <fun>
+|}]
+
+let f (local_ x : gfoo) =
+  match x with
+  | GFoo (s', _) -> ref s'
+
+[%%expect{|
+val f : local_ gfoo -> string ref = <fun>
+|}]
+
+(* the argument not marked global remains contingent on construction  *)
+(* local gives local *)
+let f (s : string) =
+  let local_ x = GFoo ("bar", s) in
+  match x with
+  | GFoo (_, s') -> s'
+
+[%%expect{|
+Line 4, characters 20-22:
+4 |   | GFoo (_, s') -> s'
+                        ^^
+Error: This local value escapes its region
+  Hint: Cannot return local value without an explicit "local_" annotation
+|}]
+
+(* and regional gives regional *)
+let f (local_ x : gfoo) =
+  match x with
+  | GFoo (_, s') -> ref s'
+
+[%%expect{|
+Line 3, characters 24-26:
+3 |   | GFoo (_, s') -> ref s'
+                            ^^
+Error: This value escapes its region
+|}]
+
+(* TESTING NONLOCAL_ *)
+(* nonlocal argument must be outer than the construction*)
+(* below, local is not outer than local*)
+let f (local_ s : string) =
+  let local_ s' = "foo" in
+  let local_ x = RFoo(s', "bar") in
+  "bar"
+[%%expect{|
+Line 3, characters 22-24:
+3 |   let local_ x = RFoo(s', "bar") in
+                          ^^
+Error: This local value escapes its region
+|}]
+
+(* local is not outer than global either *)
+let f =
+  let local_ s = "foo" in
+  let x = RFoo(s, "bar") in
+  "bar"
+[%%expect{|
+Line 3, characters 15-16:
+3 |   let x = RFoo(s, "bar") in
+                   ^
+Error: This local value escapes its region
+|}]
+
+
+(* but global is outer than local *)
+let f (s : string) =  (* s is global *)
+  let local_ x = RFoo (s, "bar") in  (* x is local *)
+  "foo"
+[%%expect{|
+Line 2, characters 13-14:
+2 |   let local_ x = RFoo (s, "bar") in  (* x is local *)
+                 ^
+Warning 26 [unused-var]: unused variable x.
```

Comment:
You should probably use `_` instead of `x` to avoid this warning

Reply:
`let local_ _ = ...` would curiously trigger an error pointing to the underscore, saying 'Syntax Error`